File: crates/oasis-brain/src/tool/knowledge.rs
use std::sync::Arc;

use async_trait::async_trait;
use oasis_core::error::OasisError;
use oasis_core::types::ToolDefinition;
use serde_json::json;

use crate::service::llm::Embedder;
use crate::service::store::VectorStore;
use crate::tool::{Tool, ToolResult};

pub struct KnowledgeTool {
    store: Arc<VectorStore>,
    embedder: Arc<Embedder>,
    vector_top_k: usize,
}

impl KnowledgeTool {
    pub fn new(store: Arc<VectorStore>, embedder: Arc<Embedder>, vector_top_k: usize) -> Self {
        Self { store, embedder, vector_top_k }
    }
}

#[async_trait]
impl Tool for KnowledgeTool {
    fn definitions(&self) -> Vec<ToolDefinition> {
        vec![ToolDefinition {
            name: "knowledge_search".to_string(),
            description: "Search the user's personal knowledge base for previously saved information, documents, and past conversations.".to_string(),
            parameters: json!({
                "type": "object",
                "properties": {
                    "query": { "type": "string", "description": "Search query" }
                },
                "required": ["query"]
            }),
        }]
    }

    async fn execute(&self, _name: &str, args: &serde_json::Value) -> ToolResult {
        let query = args["query"].as_str().unwrap_or("");
        match self.execute_search(query).await {
            Ok(r) => ToolResult::ok(r),
            Err(e) => ToolResult::err(e.to_string()),
        }
    }
}

impl KnowledgeTool {
    async fn execute_search(&self, query: &str) -> oasis_core::error::Result<String> {
        let query_embedding = self.embedder.embed(&[query]).await?;
        let embedding = query_embedding.first().ok_or_else(|| {
            OasisError::Embedding("no embedding returned".to_string())
        })?;

        let chunks = self
            .store
            .vector_search_chunks(embedding, self.vector_top_k)
            .await?;
        let relevant_messages = self.store.vector_search_messages(embedding, 5).await?;

        let mut output = String::new();
        if !chunks.is_empty() {
            output.push_str("From knowledge base:\n");
            for (i, chunk) in chunks.iter().enumerate() {
                output.push_str(&format!("{}. {}\n\n", i + 1, chunk.content));
            }
        }
        if !relevant_messages.is_empty() {
            output.push_str("From past conversations:\n");
            for msg in &relevant_messages {
                output.push_str(&format!("[{}]: {}\n", msg.role, msg.content));
            }
        }
        if output.is_empty() {
            output = format!("No relevant information found for \"{query}\".");
        }

        Ok(output)
    }
}

File: tools/knowledge/knowledge.go
package knowledge

import (
	"context"
	"encoding/json"
	"fmt"
	"strings"

	"github.com/nevindra/oasis"
)

// KnowledgeTool searches the knowledge base and past conversations.
//
// By default, New creates a HybridRetriever internally with default settings.
// To configure retrieval behavior (score threshold, filters, keyword weight,
// re-ranking), construct a Retriever with the options you need and inject it:
//
//	retriever := oasis.NewHybridRetriever(store, embedding,
//	    oasis.WithMinRetrievalScore(0.05),
//	    oasis.WithKeywordWeight(0.4),
//	    oasis.WithFilters(oasis.ByDocument("doc-123")),
//	    oasis.WithReranker(oasis.NewScoreReranker(0.1)),
//	)
//	tool := knowledge.New(store, embedding,
//	    knowledge.WithRetriever(retriever),
//	    knowledge.WithTopK(10),
//	)
type KnowledgeTool struct {
	retriever oasis.Retriever
	store     oasis.Store
	embedding oasis.EmbeddingProvider
	topK      int
}

// Option configures a KnowledgeTool.
type Option func(*KnowledgeTool)

// WithRetriever injects a custom Retriever. When not set, New creates a
// default HybridRetriever from the provided store and embedding provider.
func WithRetriever(r oasis.Retriever) Option {
	return func(k *KnowledgeTool) { k.retriever = r }
}

// WithTopK sets the number of results to retrieve. Default is 5.
func WithTopK(n int) Option {
	return func(k *KnowledgeTool) { k.topK = n }
}

// New creates a KnowledgeTool. If no Retriever is provided via WithRetriever,
// a default HybridRetriever is created from store and embedding.
func New(store oasis.Store, emb oasis.EmbeddingProvider, opts ...Option) *KnowledgeTool {
	k := &KnowledgeTool{store: store, embedding: emb, topK: 5}
	for _, o := range opts {
		o(k)
	}
	if k.retriever == nil {
		k.retriever = oasis.NewHybridRetriever(store, emb)
	}
	return k
}

func (k *KnowledgeTool) Definitions() []oasis.ToolDefinition {
	return []oasis.ToolDefinition{{
		Name:        "knowledge_search",
		Description: "Search the user's personal knowledge base for previously saved information, documents, and past conversations.",
		Parameters:  json.RawMessage(`{"type":"object","properties":{"query":{"type":"string","description":"Search query"}},"required":["query"]}`),
	}}
}

func (k *KnowledgeTool) Execute(ctx context.Context, _ string, args json.RawMessage) (oasis.ToolResult, error) {
	var params struct {
		Query string `json:"query"`
	}
	if err := json.Unmarshal(args, &params); err != nil {
		return oasis.ToolResult{Error: "invalid args: " + err.Error()}, nil
	}

	// Retrieve from knowledge base via Retriever
	chunks, err := k.retriever.Retrieve(ctx, params.Query, k.topK)
	if err != nil {
		return oasis.ToolResult{Error: "retrieval error: " + err.Error()}, nil
	}

	// Search past conversations
	embs, err := k.embedding.Embed(ctx, []string{params.Query})
	if err != nil {
		return oasis.ToolResult{Error: "embedding error: " + err.Error()}, nil
	}
	var messages []oasis.ScoredMessage
	if len(embs) > 0 {
		messages, err = k.store.SearchMessages(ctx, embs[0], 5)
		if err != nil {
			return oasis.ToolResult{Error: "message search error: " + err.Error()}, nil
		}
	}

	var out strings.Builder
	if len(chunks) > 0 {
		out.WriteString("From knowledge base:\n")
		for i, r := range chunks {
			fmt.Fprintf(&out, "%d. %s\n", i+1, r.Content)
			for _, ec := range r.GraphContext {
				fmt.Fprintf(&out, "   ↳ Related: %q (%s)\n", ec.Description, ec.Relation)
			}
			out.WriteString("\n")
		}
	}
	if len(messages) > 0 {
		out.WriteString("From past conversations:\n")
		for _, sm := range messages {
			fmt.Fprintf(&out, "[%s]: %s\n", sm.Role, sm.Content)
		}
	}
	if out.Len() == 0 {
		fmt.Fprintf(&out, "No relevant information found for %q.", params.Query)
	}

	return oasis.ToolResult{Content: out.String()}, nil
}
